input/sdl_gamepad.c

```diff
@@ -0,0 +1,244 @@
+    return INVALID_KEY;
+}
+
+static int lookup_analog_mp_key(int sdl_key, int16_t value) {
+    const int sdl_axis_max = 32767;
+    const int negative = 1;
+    const int negative_neutral = 2;
+    const int positive_neutral = 3;
+    const int positive = 4;
+
+    const float activation_threshold = sdl_axis_max * 0.33;
+    const float noise_threshold = sdl_axis_max * 0.06;
+
+    // sometimes SDL just keeps shitting out low values around 0 that mess
+    // with key repeating code
+    if (value < noise_threshold && value > -noise_threshold) {
+        return INVALID_KEY;
+    }
+
+    int state = value > 0 ? positive_neutral : negative_neutral;
+
+    if (value >= sdl_axis_max - activation_threshold) {
+        state = positive;
+    }
+
+    if (value <= activation_threshold - sdl_axis_max) {
+        state = negative;
+    }
+
+    for (int i = 0; i < MP_ARRAY_SIZE(analog_map); i++) {
+        if (analog_map[i][0] == sdl_key) {
+            return analog_map[i][state];
+        }
+    }
+
+    return INVALID_KEY;
+}
+
+
+static void request_cancel(struct mp_input_src *src)
+{
+    struct priv *p = src->priv;
+
+    MP_VERBOSE(src, "exiting...\n");
+    atomic_store(&p->cancel_requested, true);
+}
+
+static void uninit(struct mp_input_src *src)
+{
+    MP_VERBOSE(src, "exited.\n");
+}
+
+#define GUID_LEN 33
+
+static void read_gamepad_thread(struct mp_input_src *src, void *param)
+{
+    if (SDL_InitSubSystem(SDL_INIT_GAMECONTROLLER)) {
+        MP_ERR(src, "SDL_Init failed\n");
+        mp_input_src_init_done(src);
+        return;
+    };
+
+    SDL_GameController *controller;
+    struct priv *p = talloc_zero(src, struct priv);
+    char guid[GUID_LEN];
+
+    if (SDL_NumJoysticks() <= 0) {
+        MP_VERBOSE(src, "no joysticks found");
+        mp_input_src_init_done(src);
+        return;
+    }
+
+    MP_VERBOSE(src, "connected controllers: %i\n", SDL_NumJoysticks());
+
+    for (int i = 0; i < SDL_NumJoysticks(); ++i) {
+        if (SDL_IsGameController(i)) {
+            controller = SDL_GameControllerOpen(i);
+            SDL_JoystickID id = SDL_JoystickInstanceID(
+                SDL_GameControllerGetJoystick(controller));
+
+            SDL_JoystickGetGUIDString(
+                SDL_JoystickGetGUID(SDL_GameControllerGetJoystick(controller)),
+                guid, GUID_LEN);
+
+            if (controller) {
+                MP_VERBOSE(
+                    src, "detected controller #%i: %s, guid: %s\n",
+                    id, SDL_GameControllerName(controller), guid);
+
+                // stop at first controller apparently SDL can't open more
+                // than one controller anyway?
+                break;
+            }
+        }
+    }
+
+    atomic_store(&p->cancel_requested, false);
+    src->priv = p;
+    src->cancel = request_cancel;
+    src->uninit = uninit;
+    mp_input_src_init_done(src);
+
+    SDL_Event ev;
+
+    while (!atomic_load(&p->cancel_requested)) {
+        while (SDL_PollEvent(&ev) != 0) {
```

Comment:
It looks like this will just burn a core continuously.

Reply:
Thanks, fixed. That's why my room was awfully hot when using mpv in the past months.